spi/src/main/java/org/jboss/pnc/spi/datastore/repositories/BuildRecordRepository.java

```diff
@@ -48,20 +48,22 @@
     BuildRecord getLatestSuccessfulBuildRecord(Integer configurationId, boolean buildTemporary);
 
     default BuildRecord getLatestSuccessfulBuildRecord(List<BuildRecord> buildRecords, boolean buildTemporary) {
+        final boolean containsTemporary = buildRecords.stream().anyMatch(BuildRecord::isTemporaryBuild);
+
+        //Include temporary builds if you are building temporary and don't if building persistent
+        final boolean includeTemporary = buildTemporary;
+        //NCL-5192
+        //Exclude persistent builds if you are building temporary and there are some temporary builds built
+        final boolean excludePersistent = buildTemporary && containsTemporary;
+        final boolean includePersistent = !excludePersistent;
+
         return buildRecords.stream()
                 .filter(record -> record.getStatus() == BuildStatus.SUCCESS)
-                /*
-                 * filter out the temporary records if we are building persistent(temporaryBuild == false)
-                 * For clarification:
-                 *  |     temporaryBuild     | !record.isTemporaryBuild() |  CONDITION IN FILTER  |
-                 *  |         TRUE           |          TRUE              |        TRUE           |
-                 *  |         TRUE           |          FALSE             |        TRUE           |
-                 *  |         FALSE          |          TRUE              |        TRUE           |
-                 *  |         FALSE          |          FALSE             |        FALSE          |
-                 */
-                .filter(record -> buildTemporary || !record.isTemporaryBuild() )
-                .sorted((o1, o2) -> -o1.getId().compareTo(o2.getId()))
-                .findFirst().orElse(null);
+                //First part includes temporary BRs and second part includes persitent BRs
+                .filter(record -> (includeTemporary && record.isTemporaryBuild()) ||
+                                  (includePersistent && !record.isTemporaryBuild()))
```

Comment:
What about including the one or another filter based on the condition? It would make the filters simpler and more readable.

Reply:
On what condition? The `buildTemporary`? I think that would make it simple only in one (persistent) branch.

Reply:
Hmm, it would look like this:
```java
Predicate<BuildRecord> filter;
if(buildTemporary){
    final boolean containsTemporary = buildRecords.stream().anyMatch(BuildRecord::isTemporaryBuild);
    final boolean includePersistent = !containsTemporary;
    filter = record -> record.isTemporaryBuild() || includePersistent;
}else{
    filter = record -> !record.isTemporaryBuild();
}
```

Reply:
This is even harder to read.